Source/Core/Common/TraversalServer.cpp

```diff
@@ -0,0 +1,458 @@
+
+			if (info->packet.type == TraversalPacketPleaseSendPacket)
+			{
+				TraversalPacket* ready = AllocPacket(MakeSinAddr(info->packet.pleaseSendPacket.address));
+				if (packet->ack.ok)
+				{
+					ready->type = TraversalPacketConnectReady;
+					ready->connectReady.requestId = info->misc;
+					ready->connectReady.address = MakeInetAddress(info->dest);
+				}
+				else
+				{
+					ready->type = TraversalPacketConnectFailed;
+					ready->connectFailed.requestId = info->misc;
+					ready->connectFailed.reason = TraversalConnectFailedClientFailure;
+				}
+			}
+
+		outgoingPackets.erase(it);
+		break;
+		}
+	case TraversalPacketPing:
+		{
+		auto r = EvictFind(connectedClients, packet->ping.hostId, true);
+		packetOk = r.found;
+		break;
+		}
+	case TraversalPacketHelloFromClient:
+		{
+		u8 ok = packet->helloFromClient.protoVersion <= TraversalProtoVersion;
+		TraversalPacket* reply = AllocPacket(*addr);
+		reply->type = TraversalPacketHelloFromServer;
+		reply->helloFromServer.ok = ok;
+		if (ok)
+		{
+			TraversalHostId hostId;
+			TraversalInetAddress* iaddr;
+			// not that there is any significant change of
+			// duplication, but...
+			GetRandomHostId(&hostId);
+			while (1)
+			{
+				auto r = EvictFind(connectedClients, hostId);
+				if (!r.found)
+				{
+					iaddr = EvictSet(connectedClients, hostId);
+					break;
+				}
+			}
+
+			*iaddr = MakeInetAddress(*addr);
+
+			reply->helloFromServer.yourAddress = *iaddr;
+			reply->helloFromServer.yourHostId = hostId;
+		}
+		break;
+		}
+	case TraversalPacketConnectPlease:
+		{
+		TraversalHostId& hostId = packet->connectPlease.hostId;
+		auto r = EvictFind(connectedClients, hostId);
+		if (!r.found)
+		{
+			TraversalPacket* reply = AllocPacket(*addr);
+			reply->type = TraversalPacketConnectFailed;
+			reply->connectFailed.requestId = packet->requestId;
+			reply->connectFailed.reason = TraversalConnectFailedNoSuchClient;
+		}
+		else
+		{
+			TraversalPacket* please = AllocPacket(MakeSinAddr(*r.value), packet->requestId);
+			please->type = TraversalPacketPleaseSendPacket;
+			please->pleaseSendPacket.address = MakeInetAddress(*addr);
+		}
+		break;
+		}
+	default:
+		fprintf(stderr, "received unknown packet type %d from %s\n", packet->type, SenderName(addr));
+	}
+	if (packet->type != TraversalPacketAck)
+	{
+		TraversalPacket ack = {0};
+		ack.type = TraversalPacketAck;
+		ack.requestId = packet->requestId;
+		ack.ack.ok = packetOk;
+		TrySend(&ack, sizeof(ack), addr);
+	}
+}
+
+int main()
+{
+	int rv;
+
+	urandomFd = open("/dev/urandom", O_RDONLY);
+	if (urandomFd < 0)
+	{
+		perror("open /dev/urandom");
+		return 1;
+	}
+
+	sock = socket(PF_INET6, SOCK_DGRAM, 0);
+	if (sock == -1)
+	{
+		perror("socket");
+		return 1;
+	}
+	int no = 0;
+	rv = setsockopt(sock, IPPROTO_IPV6, IPV6_V6ONLY, &no, sizeof(no));
+	if (rv < 0)
+	{
+		perror("setsockopt IPV6_V6ONLY");
+		return 1;
+	}
+	struct in6_addr any = IN6ADDR_ANY_INIT;
+	struct sockaddr_in6 addr;
```

Comment:
You never need typedef/struct... in C++ because all structs/unions/classes/enums are implicitly typedefed.